File: docs/l03-pid-control/code/q6_starter.py
import control as ct
import numpy as np

s = ct.tf("s")
G = 40 / (s * (s + 2) * (s + 10))        # pitch attitude / elevator (rad/rad)
Kp, Ki, Kd, N = 2.03, 0.678, 0.661, 10   # the lecture's PID design
Tf = Kd / (N * Kp)

C = Kp + Ki / s + Kd * s / (Tf * s + 1)  # full controller: sets the loop
Cr = Kp + Ki / s                         # the reference sees P and I only
T_ref = ct.minreal(Cr * G / (1 + C * G), verbose=False)   # reference to attitude
T_u = ct.minreal(Cr / (1 + C * G), verbose=False)         # reference to elevator

gm, pm, w180, wc = ct.margin(C * G)
print(f"PM = {pm:.1f} deg at {wc:.2f} rad/s")

t = np.linspace(0, 10, 5001)
u = ct.step_response(T_u, t).outputs * 10   # elevator (deg) for a 10 deg step
print(f"peak elevator = {np.max(np.abs(u)):.1f} deg")

# (b) Scale Kp, Ki and Kd by the same factor k and search for the largest k
#     meeting the requirements. Your code here.

File: docs/w03-pid-control/code/q6_starter.m
s = tf('s');
G = 40 / (s*(s + 2)*(s + 10));   % pitch attitude / elevator (rad/rad)
Kp = 2.03; Ki = 0.678; Kd = 0.661; N = 10;   % the lecture's PID design

C = pid(Kp, Ki, Kd, Kd/(N*Kp));  % full controller: sets the loop
Cr = pid(Kp, Ki);                % the reference sees P and I only
Tref = Cr * feedback(G, C);      % reference to attitude
Tu = Cr * feedback(1, C*G);      % reference to elevator

[~, Pm, ~, Wcp] = margin(C*G);
fprintf('PM = %.1f deg at %.2f rad/s\n', Pm, Wcp)

t = linspace(0, 10, 5001);
u = step(Tu, t) * 10;            % elevator (deg) for a 10 deg step
fprintf('peak elevator = %.1f deg\n', max(abs(u)))

% (b) Scale Kp, Ki and Kd by the same factor k and search for the largest k
%     meeting the requirements. Your code here.

% tracking: status=draft version=0 assisted=true
